docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,1063 @@
+     * [Test Plan](#test-plan)
+     * [Version Skew Strategy](#version-skew-strategy)
+  * [Implementation History](#implementation-history)
+  * [Controller Runtime Types](#controller-runtime-types)
+
+## Glossary
+
+The lexicon used in this document is described in more detail
+[here](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/reference/glossary.md).
+Any discrepancies should be rectified in the main Cluster API glossary.
+
+## Summary
+
+The clusterctl CLI currently handles the lifecycle of Cluster API
+providers installed in a management cluster. It provides a great Day 0 and Day
+1 experience in getting CAPI up and running. However, clusterctl’s imperative
+design makes it difficult for cluster admins to stand up and manage CAPI
+management clusters in their own preferred way.
+
+This proposal provides a solution that leverages a declarative API and an
+operator to empower admins to handle the lifecycle of providers within the
+management cluster.
+
+## Motivation
+
+In its current form clusterctl is designed to provide a simple user experience
+for day 1 operations of a Cluster API management cluster.
+
+However such design is not optimized for supporting declarative approaches
+when operating Cluster API management clusters.
+
+These declarative approaches are important to enable GitOps workflows in case
+users don't want to rely solely on the `clusterctl` CLI.
+
+Providing a declarative API also enables us to leverage controller-runtime's
+new component config and allow us to configure the controller manager and even
+the resource limits of the provider's deployment.
+
+Another example is improving cluster upgrades. In order to upgrade a cluster
+we now need to supply all the information that was provided initially during a
+`clusterctl init` which is inconvenient in many cases such as distributed
+teams and CI pipelines where the configuration needs to be stored and synced
+externally.
+
+With the management cluster operator, we aim to address these use cases by
+introducing an operator that handles the lifecycle of providers within the
+management cluster based on a declarative API.
+
+### Goals
+
+- Define an API that enables declarative management of the lifecycle of
+  Cluster API and all of its providers.
+- Support air-gapped environments through sufficient documentation initially.
+- Identify and document differences between clusterctl CLI and the operator in
+  managing the lifecycle of providers, if any.
+- Define how the clusterctl CLI should be changed in order to interact with
+  the management cluster operator in a transparent and effective way.
+- To support the ability to upgrade from a v1alpha3 based version (v0.3.[TBD])
+  of Cluster API to one managed by the operator.
+
+### Non-Goals/Future Work
+
+- Implement an operator driven version of `clusterctl move`.
+- Manage cert-manager using the operator.
+- Support multiple installations of the same provider within a management
```

Comment:
Coming from the discussion yesterday again:
Not supporting it here should be fine but would it still be possible?
As far as I can tell, there is nothing directly prohibiting me from having 2 installations of the same provider in different namespaces here.

Reply:
That's correct, the operator itself doesn't stop any user to be able to run multiple controllers in a management cluster, but it doesn't/won't support it natively for the v1alpha4 iteration.

The Cluster API community has been going in this direction because it's a hard problem to support version skews in the same management cluster with a generic operator. The goal of this operator proposal is to fulfill the 80% use case: run a management cluster with a simple deployment. Meanwhile, we should make the 20% possible (same as every other feature we usually build).

Reply:
We may need to revise these statements in the _Installing a Provider_ section

https://github.com/kubernetes-sigs/cluster-api/pull/3833/files#diff-a8b3fa322e4449932bb3e414c2582ca1acae4ae38e9fd9036e3df433f5b5528bR612-R615

> Before installing any provider following preflight checks will be executed :
>- There should not be another instance of the same provider (same Kind, same name) in any namespace.
>- The Cluster API contract the provider is abiding by, e.g. v1alpha4, must match the contract of the core provider.

Reply:
I think we should keep these checks in the operator because it should ensure the consistency of the cluster no matter if it is used with clusterctl or not.

Reply:
The checks should still be in place, but we shouldn't disallow running more than one controller in Cluster API code. The difference here is that the use cases supported by the operator are on a smaller scale, while we allow others to run different deployments if they need to.